astropy/utils/data.py

```diff
@@ -952,7 +952,7 @@ def get_pkg_data_fileobjs(datadir, package=None, pattern="*", encoding=None):
 
 
 def compute_hash(localfn):
-    """Computes the MD5 hash for a file.
+    """Computes the SHA256 hash for a file.
```

Comment:
This is not backward compatible. I think we need to give user an option. But not sure how to deprecate MD5 in a not painful way.

Reply:
Maybe it is less controversial to introduce SHA256 as the non-default option in this PR.

Reply:
That could work. Probably the most robust way to do this would be something like:
```
from cryptography.hazmat.primitives.fips import is_fips_mode_enabled

if is_fips_mode_enabled():
    hashlib.sha256...
else:
    hashlib.md5...
```
However, this introduces the cryptography and openssl packages as dependencies, and is probably totally overkill.

Another option might be an environment variable:
```
import os

if os.environ.get('ASTROPY_FIPS', '0') == '1':
    hashlib.sha256...
else:
    hashlib.md5...
```

Perhaps the most elegant way might be to simply do:
```
try:
    hashlib.md5...
except ValueError:
    hashlib.sha256...
```

Reply:
I think your third proposition comes closest to how astropy idiomatically handles optional dependencies. However, I would still suggest to avoid a try/except where possible, something like
```python
hash_func = getattr(hashlib, "md5", hashlib.sha256)
```

Reply:
`getattr` doesn't work here unfortunately as md5 is an available attribute and can be retrieved as such successfully. It only throws an exception if you try to use it:
```
>>> import hashlib
>>> hash_func = getattr(hashlib, "md5", hashlib.sha256)
>>> print(hash_func)
<built-in function openssl_md5>
>>> print('md5' in hashlib.algorithms_guaranteed)
True
>>> hash_func().update(b"test")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: [digital envelope routines] unsupported
```
(See the [hashlib docs](https://docs.python.org/3/library/hashlib.html) regarding Python's questionable decision to include md5 in algorithms_guaranteed despite FIPS issues).

Perhaps a simple wrapper class to "hide the sins" so-to-speak might be desirable:
```
import hashlib

class _Hasher:
    def __init__(self):
        self._hash_method = 'MD5'
        try:
            hashlib.md5().update(b"Test MD5 availability")
        except ValueError:
            self._hash_method = 'SHA256'
            # Maybe log something here to notify the user
    
    def get_hash_func(self):
        if self._hash_method == 'MD5':
            return hashlib.md5()
        else:
            return hashlib.sha256()

hasher = _Hasher()

# Replace instances of 
# >>> import hashlib
# >>> x = hashlib.md5()
# with
# >>> from hasher import hasher
# >>> x = hasher.get_hash_func()
```

Reply:
Wait a minute! This is new:
```
Changed in version 3.9: All hashlib constructors take a keyword-only argument usedforsecurity with default value True. A false value allows the use of insecure and blocked hashing algorithms in restricted environments. False indicates that the hashing algorithm is not used in a security context, e.g. as a non-cryptographic one-way compression function.
```

Since the latest versions of astropy only support down to 3.10 (correct me if I'm wrong), and the md5 hash isn't used for security purposes, I think it'd be appropriate to simply replace instances of `hashlib.md5()` with `hashlib.md5(usedforsecurity=False)`. This appears to work in my test environment:
```
>>> hash_func = hashlib.md5(usedforsecurity=False)
>>> hash_func.update(b"Test")
# No exception
```
Thoughts about this? The only thing would be looking out for future usages of hashlib.md5 without inclusion of the usedforsecurity flag.

Reply:
> the latest versions of astropy only support down to 3.10 (correct me if I'm wrong)

you are correct !

> Thoughts about this?

This seems both appropriate and practical to me !

> The only thing would be looking out for future usages of hashlib.md5 without inclusion of the usedforsecurity flag.

Excellent idea, but it feels like a job for a linter. I would suggest opening an issue to `ruff` and proposing this as a rule !

Reply:
👍 to utilize `usedforsecurity=False` and, if you feel strongly about this topic, please open follow-up issue about maybe switching to SHA256 in the future. Thanks!

Reply:
Implemented in 6c84035